From 71de1e422175e7819d7f2c5dac76879b4e6f549c Mon Sep 17 00:00:00 2001
From: "no-reply@wasmlabs.dev" <no-reply@wasmlabs.dev>
Date: Fri, 6 Jan 2023 11:21:03 +0200
Subject: [PATCH 4/9] fix: Fix memory issues and introduce WASM_WASMEDGE.

  76.4% Zend/
  11.4% ext/posix/
  12.1% ext/standard/

  76.4% Zend/
  11.4% ext/posix/
  12.1% ext/standard/
diff --git a/Zend/zend_alloc.c b/Zend/zend_alloc.c
index 5e67345c..258766a5 100644
--- a/Zend/zend_alloc.c
+++ b/Zend/zend_alloc.c
@@ -465,8 +465,8 @@ static void *zend_mm_mmap(size_t size)
 	}
 	return ptr;
 #elif ! HAVE_MMAP
-	void* ptr = malloc(size);
-	memset(ptr, 0, size);
+	void* ptr = malloc(size + ZEND_MMAP_AHEAD);
+	memset(ptr, 0, size + ZEND_MMAP_AHEAD);
 	return ptr;
 #else
 	void *ptr;
@@ -697,8 +697,8 @@ static zend_always_inline void zend_mm_hugepage(void* ptr, size_t size)
 static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 {
 #if ! HAVE_MMAP
-	void* ptr = aligned_alloc(alignment, size);
-	memset(ptr, 0, size);
+	void* ptr = aligned_alloc(alignment, size + ZEND_MMAP_AHEAD);
+	memset(ptr, 0, size + ZEND_MMAP_AHEAD);
 	return ptr;
 #else
 	void *ptr = zend_mm_mmap(size);
diff --git a/ext/posix/posix.c b/ext/posix/posix.c
index 49806f2c..c5e5c4a0 100644
--- a/ext/posix/posix.c
+++ b/ext/posix/posix.c
@@ -536,8 +536,12 @@ PHP_FUNCTION(posix_ttyname)
 		RETURN_FALSE;
 	}
 #endif
+#else
+	p = emalloc(1);
+	*p = '\0';
 #endif // WASM_WASI
 	RETURN_STRING("");
+	efree(p);
 }
 /* }}} */
 
diff --git a/ext/standard/basic_functions.c b/ext/standard/basic_functions.c
index e3e443b4..d367dd8e 100755
--- a/ext/standard/basic_functions.c
+++ b/ext/standard/basic_functions.c
@@ -70,6 +70,8 @@ typedef struct yy_buffer_state *YY_BUFFER_STATE;
 #else
 #include "win32/inet.h"
 #endif
+#elif defined(WASM_WASMEDGE)
+# include <netdb.h>
 #endif // WASM_WASI
 
 #ifdef HAVE_ARPA_INET_H
-- 
2.38.1

